### ISSUE TITLE
Feature request, node temp sensor

### ISSUE THREAD

**Issue by jrstrengen** [mrgadget/node-red-contrib-hdlbus#5]
I found out that it could be nifty to work with external temp sensors for my HDL env.
I use a 433 mhz sensor in homeassistant and update my enviropanel with outdoor temp read from the 433 sensor.

I use the HDL raw inputs/outputs. But this adds switches to filter out only the packages i need. 
Could it be an idea to add a node that reads a temp request  and a node that generates a temp response? Basically a virtual temp sensor. This node could potentially have its own id on the bus?

What I have works, but I'm thinking on how to make the plugin more userfriendly.

Messge received from HDL requesting outdoor temp:
```
sender: "1.100"
target: "254.99"
code: 6472
payload: object
topic: "command"
```

Messge sent from HDL updating outdoor temp
```
target: "1.100"
sender: "254.99"
code: "6473"
payload: object
channel: 1
temperature: "-3.9"
```

Messages on the bus
```
00001  09:15:21:968 : 0C 01 65 00 B9 19 48 FE 63 01 8B 67     
00002  09:15:21:970 : 10 FE 63 FF FE 19 49 01 65 01 CD CC 6C C0 E5 32     :-3,7   254-99--Unname  ---> 1-101-panel
```

code for indoor temp request looks like it is: dec(58343) hex(e3e7)and response dec(58344),hex(e3e8)

**Comment by mrgadget** [comment]
Sounds like a reasonable request - I see this as a node with only an input (to set the temperature that it will return) and a configuration option for virtual address - which it will list on and respond from.  Am I understanding correctly?

I already do something similar with a virtual device to 'fix' dali groups, so I could re-purpose that.

**Comment by mrgadget** [comment]
Have a shot at that - replace the the files in the `hdlbus` folder with the ones from this repo and restart node-red.  The new node should appear and work as above.

In the snip below I hit the input node once just to inject/set the temperature (no log/output occurs).  I then sent a `0x1948` (6472) request from the HDL network, which the node picked up and responded to with a `0x1949` (6473) response - both of which are detailed in the debug pane on the right.

Thoughts?

![image](https://user-images.githubusercontent.com/700522/75542142-13a53b00-5a84-11ea-836e-d48bdcc347d8.png)

**Comment by jrstrengen** [comment]
The node itself do as designed. But the enviropanel won't pick up the new temperature. However the idea of the solution is perfect.

I see it picks up the request, and generates the correct response and broadcast it.
The panel is a MPTLC43.46

Virtual ID: was set to 20.10.1 -> this gets picked up
Response is stored from temperature session.

Response source is 254.99 (Node-red controller ID)
Response target is FF FF (Broadcast)

`
00007  16:26:21:608 : 0C 01 65 00 B9 19 48 14 0A 01 5D 25     
00008  16:26:21:610 : 10 14 0A FF FE 19 49 FF FF 01 00 00 20 41 48 81     :10   20-10--Unname
`

I also tested setting Virtual ID to same as Response to match these, the same target adress is used.
I expect that we need to return the Response to the original source device.
Also i do not think it will accept a temp update from another source than specified in the request.
This is the only differences i find in the packet made by my raw device and the temp device.

Since we use 1948/1949 as codes, this will also only apply to outdoor temp setting on the panel.
Should have an option to select outdoor/indoor to change code.

indoor is 58343/58344

**Comment by jrstrengen** [comment]
This is how i solved it initally:
![image](https://user-images.githubusercontent.com/59741565/75563327-7a0d6780-5a4a-11ea-90a8-2d5673efa818.png)

The HA outdoor temp sets a global variable that i picked up in the  create response

`
orgmsg =this.msg
newmsg= {
    target:orgmsg.sender,
    sender:orgmsg.target,
    code:6473,
    payload:{
        channel: orgmsg.payload.channel,
        //temperature: '-5'
        temperature: flow.get("outdoor")
        //level: orgmsg.payload || '150'
    }
}
return newmsg;
`

**Comment by mrgadget** [comment]
Understand - just uploaded a new version, give that a go.

I've adjusted the HDL response to use the source of the request as the target of the response (instead of broadcast).

After looking at `0xE3E7` and comparing the request/response payloads, they look the same - so I've just made the node respond to either.   If you send if a `0x1948` request, it'll respond with a `0x1949`, conversely, if you send if a `0xE3E7`, it'll respond with a `0xE3E8`.  I think that sorts this one out?

**Comment by jrstrengen** [comment]
Thats the easy way to fix it. 
The package is identical, its just the command code that is different. 

I will test the new version later today, and give feedback.

**Comment by jrstrengen** [comment]
I've tested new version.
It looks good. I see package comes from the HDL controller node and are sent to the node that requests.
Looks good for both virtual ID and HDL controller node.

Now my next part is to integrate  lights into home-assistant and be able to control and see status on both wall and app.

**Comment by mrgadget** [comment]
Yep - done that (pretty much the driver for me working on this module).

I chose MQTT - I was new to HA at the time, maybe there's a better way (using HA integration), but it works for me.

I have this _Node-Red_ flow;
![image](https://user-images.githubusercontent.com/700522/75613381-dcbd4b00-5b91-11ea-8cda-7bb0003bbf9e.png)

You only need this flow once - it has variables for input and output so it will work for any address on your network.  

Here's the JSON so you can import it.  It will create new config nodes for MQTT and HDL (with my ip details etc) - you'll need to go into the two MQTT nodes and the two HDL nodes and change them to your config node (then delete the newly created config nodes).
```json
[{"id":"b32157c7.433488","type":"comment","z":"d2856ab9.5a3dd8","name":"lights","info":"","x":140,"y":645,"wires":[]},{"id":"4c2ba2a8.b13b5c","type":"mqtt in","z":"d2856ab9.5a3dd8","name":"from hass.io","topic":"lights/+/+/cmd","qos":"0","datatype":"auto","broker":"3ae6510.9fc8ab","x":325,"y":615,"wires":[["5a06debd.a6e9a","ff8c9a01.0c9548"]]},{"id":"ff8c9a01.0c9548","type":"function","z":"d2856ab9.5a3dd8","name":"hass > hdl","func":"if (msg.payload == \"ON\") {\n    msg.payload = 100;\n} else if (msg.payload == \"OFF\") {\n    msg.payload = 0;\n}\n    \nmsg.payload = {\n    \"address\": \"1.\" + msg.topic.split(\"/\")[1],\n    \"channel\": parseInt(msg.topic.split(\"/\")[2]),\n    \"level\": parseInt(msg.payload)\n}\nreturn msg;","outputs":1,"noerr":0,"x":640,"y":615,"wires":[["4654791.db8ed88","5a66c3d4.28de6c"]]},{"id":"5a66c3d4.28de6c","type":"hdl-channel-out","z":"d2856ab9.5a3dd8","controller":"42aa3da7.36f314","name":"to hdl","address":"1.1","channel":1,"level":"","x":890,"y":600,"wires":[]},{"id":"1586ad61.c3c5f3","type":"hdl-raw-in","z":"d2856ab9.5a3dd8","controller":"42aa3da7.36f314","name":"from hdl","x":315,"y":675,"wires":[["f1462740.2675a8"]]},{"id":"9d7789a5.920a38","type":"function","z":"d2856ab9.5a3dd8","name":"hdl > hass.io","func":"if (msg.code == 50) {\n    var out = { \n        topic: \"lights\",\n        payload: null\n    }\n    out.topic += \"/\" + msg.sender.split(\".\")[1];\n    out.topic += \"/\" + msg.payload.channel;\n    out.topic += \"/state\";\n    if (msg.payload.level === 0) {\n        out.payload = {\n            \"power\": \"OFF\",\n            \"brightness\": msg.payload.level\n        }\n    } else {\n        out.payload = {\n            \"power\": \"ON\",\n            \"brightness\": msg.payload.level\n        }\n    }\n}\nreturn out;","outputs":1,"noerr":0,"x":640,"y":675,"wires":[["4139b534.a1493c","45bf119b.5722f"]]},{"id":"45bf119b.5722f","type":"mqtt out","z":"d2856ab9.5a3dd8","name":"to hass.io","topic":"","qos":"1","retain":"true","broker":"3ae6510.9fc8ab","x":900,"y":690,"wires":[]},{"id":"4139b534.a1493c","type":"debug","z":"d2856ab9.5a3dd8","name":"hass output","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","x":910,"y":645,"wires":[]},{"id":"5a06debd.a6e9a","type":"debug","z":"d2856ab9.5a3dd8","name":"from hass.io","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","x":640,"y":570,"wires":[]},{"id":"960c051a.d0ba28","type":"debug","z":"d2856ab9.5a3dd8","name":"from hdl","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","x":630,"y":720,"wires":[]},{"id":"4654791.db8ed88","type":"debug","z":"d2856ab9.5a3dd8","name":"hdl output","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","x":900,"y":555,"wires":[]},{"id":"f1462740.2675a8","type":"function","z":"d2856ab9.5a3dd8","name":"filter","func":"if (msg.code == 50) {\n    return msg;\n}","outputs":1,"noerr":0,"x":470,"y":675,"wires":[["960c051a.d0ba28","9d7789a5.920a38"]]},{"id":"3ae6510.9fc8ab","type":"mqtt-broker","z":"","name":"home-mqtt","broker":"mqtt.local","port":"1883","clientid":"","usetls":false,"compatmode":true,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closeQos":"0","closePayload":"","willTopic":"","willQos":"0","willPayload":""},{"id":"42aa3da7.36f314","type":"hdl-controller","z":"","host":"192.168.1.255","port":"6000","subnetid":"1","deviceid":"98","broadcast":true,"daliId":"20"}]
```

Then in _Home Assistant_ `lights` section;
```yaml
#Relay
- platform: mqtt
  name: "Outside Light"
  command_topic: "lights/30/1/cmd"
  state_topic: "lights/30/1/state"
  state_value_template: "{{value_json.power}}"

#Dimmer
- platform: mqtt
  name: "Lounge Light"
  command_topic: "lights/20/1/cmd"
  state_topic: "lights/20/1/state"
  state_value_template: "{{value_json.power}}"
  on_command_type: brightness
  brightness_scale: 100
  brightness_command_topic: "lights/20/1/cmd"
  brightness_state_topic: "lights/20/1/state"
  brightness_value_template: "{{value_json.brightness}}"
```

If you find a better way to do it (eg using the Home Assistant integration nodes), please share! :)

**Comment by jrstrengen** [comment]
Thanks for the examples on how you did it.
I’m new to ha and node-red, and HDL is sometimes a mystery in itself.
With your config, ha & node red became clearer than any of the examples from official documentation.

I think I will go your way first to get it up and running. Unless I misunderstood the ha integration completely, I would most likely end up with many flows. Or it might be me over complicating things ;)

I’ll be happy to share any smart integrations. And I’ll keep following here and will be happy to test anything if needed.

Is HDL big in NZ? Here in Norway it seems like they have stopped a little. Have not seen any recent firmware updates made available. Not sure if this is due to the Norwegian partner or if there are little development of existing devices.

**Comment by mrgadget** [comment]
Totally agree re: the mystery of HDL. Its hard to get accurate information from anywhere - there are a couple of documents floating around for operation codes which give you the main ones. Some others (panel brightness and colours) I watched the network to see what was being sent. 

I think the HA integration could be done in a similar way (with variables) but I’m not 100% sure.  I struggle with motivation when this method works reliably for me!

HDL in NZ is very niche. It’s around, but not very heavily used. Unfortunately our distributor chose not to import the wireless side which I was interested in. I’ve put it in (myself) a couple of houses and helped with the setup/troubleshooting in a handful more. I like it though, and it’s reliable. The first house I built was 8 years ago and it’s never missed a beat.
